stanify/builders/stan_block_builder.py

```diff
@@ -0,0 +1,673 @@
+        self.precision_context = precision_context
+        self.stan_model_context = stan_model_context
+        self.vensim_model_context = vensim_model_context
+
+
+    def build_block(self):
+        self.code = IndentedString()
+        self.code += "generated quantities{\n"
+        self.code.indent_level += 1
+        self.build_post_pred_rng_functions()
+        self.code += "\n"
+        self.build_loglik_functions()
+        self.code.indent_level -= 1
+        self.code += "}\n"
+
+        return str(self.code)
+
+    def build_loglik_functions(self):
+        self.code += "real loglik;\n"
+
+        if self.precision_context.R == 1:
+            for statement in self.stan_model_context.sample_statements:
+                if statement.lhs_expr in self.stan_model_context.obs_integ_outcome_vector_names:
+                    param_name = statement.lhs_expr
+                    loc = statement.distribution_args[0]
+                    scale = statement.distribution_args[1]
+                    if statement.distribution_type in ["normal", "lognormal"]:
+                        self.code += f"loglik += {statement.distribution_type}_lpdf({param_name}|{loc}, {scale});\n"
+                    elif statement.distribution_type in ["neg_binom_2"]:
+                        self.code += f"loglik += {statement.distribution_type}_lpmf({param_name}|{loc}, {scale});\n"
+        else:
+            self.code += "for (r in 1:R){\n"
+            self.code.indent_level += 1
+            for statement in self.stan_model_context.sample_statements:
+                if statement.lhs_expr in self.stan_model_context.obs_integ_outcome_vector_names:
+                    obs_vec_name = statement.lhs_expr
+                    target_vec_name = obs_vec_name[:-4]
+                    scale = statement.distribution_args[1]
+                    if statement.distribution_type in ["normal", "lognormal"]:
+                        self.code += f"loglik += {statement.distribution_type}_lpdf({obs_vec_name}[:, r]|{target_vec_name}[:, r], {scale});\n"
+                    elif statement.distribution_type in ["neg_binom_2"]:
+                        self.code += f"loglik += {statement.distribution_type}_lpmf({obs_vec_name}[:, r]|{target_vec_name}[:, r], {scale});\n"
+            self.code.indent_level -= 1
+            self.code += "}\n"
+
+    def build_post_pred_rng_functions(self):
+
+        if self.precision_context.R == 1:
+            self.code += "// Define and assign generated value to posterior predictive vector\n"
+            for statement in self.stan_model_context.sample_statements:
+                if statement.lhs_expr in self.stan_model_context.obs_integ_outcome_vector_names:
+                    # TODO @Dashadower how to use the following in the future?
+                    #  stan_type = self.stan_model_context.stan_data[statement.lhs_expr].stan_type #     stan_type = self.stan_model_context.stan_data[statement.lhs_expr].stan_type KeyError: 'prey_obs'
```

Comment:
> @Dashadower how to use the following in the future?
> stan_type = self.stan_model_context.stan_data[statement.lhs_expr].stan_type 
> stan_type = self.stan_model_context.stan_data[statement.lhs_expr].stan_type KeyError: 'prey_obs'

`stan_model_context.stan_data` only has data types for **data** variables. A KeyError happening would probably indicate the variable isn't from the data block but instead is a parameter/transformed parameter. In that case, there's no way of inferring its exact type unless you pass additional information to the posterior predictive rng function builder.